### PR TITLE
ANP Context Dropout: robust context aggregation for p_re generalization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Neural Processes (Garnelo et al. 2018, Kim et al. 2019) are specifically designed to generalize across context sizes and distributions by training with **randomly subsampled context sets**. Our ANP decoder currently sees ALL surface nodes as context — always. This creates the same fragility as training without dropout: the model memorizes full-context patterns.

At OOD Reynolds numbers, the context node features shift (different pressure magnitudes), effectively presenting a "corrupted context." A model trained to always see full context has no robustness mechanism. Training with **random context masking** (drop 20-30% of context nodes per forward pass) forces the ANP to aggregate information robustly from partial context — generalizing to distribution-shifted contexts including OOD Re.

This is well-motivated by the Neural Process literature: this technique is standard in CNP/ANP training for exactly this reason (proper marginalization over context sets). We skipped this step when implementing the ANP decoder.

### Step 1: Implement `--anp_context_dropout`

In the ANP decoder's context encoding step, add training-time random masking:

```python
# During training only (model.training == True):
if self.anp_context_dropout > 0 and self.training:
    # context_nodes: [B, N_context, D]
    B, N, D = context_nodes.shape
    # Create random mask: keep each node with probability (1 - p_drop)
    mask = torch.rand(B, N, 1, device=context_nodes.device) > self.anp_context_dropout
    # Zero out dropped context nodes
    context_nodes = context_nodes * mask.float()
    # Note: we drop nodes (zero them) rather than removing them to keep attention shape fixed
    # Alternatively, use masked attention (set attention mask for dropped nodes to -inf)
```

**Masked attention approach (preferred):** Set the attention mask for dropped context positions to -inf before softmax, so the model truly ignores dropped nodes rather than seeing zeroed features:

```python
# In the attention score computation:
# scores: [B, N_query, N_context]
if dropped_mask is not None:  # dropped_mask: [B, N_context], True = dropped
    scores = scores.masked_fill(dropped_mask.unsqueeze(1), float('-inf'))
```

At **inference**: use full context (no masking). This is stochastic depth for the context set.

Add flag `--anp_context_dropout 0.3` (try both 0.2 and 0.3).

### Step 2: Training runs (use `--wandb_group anp-context-dropout`)

Run 3 configs with seed 42 first:

```bash
# Config A: p_drop=0.2
CUDA_VISIBLE_DEVICES=0,1 python train.py \
  --agent nezuko --wandb_name "nezuko/anp-ctx-drop-02-s42" \
  --wandb_group anp-context-dropout --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --anp_srf --anp_context_dropout 0.2

# Config B: p_drop=0.3
CUDA_VISIBLE_DEVICES=2,3 python train.py \
  --agent nezuko --wandb_name "nezuko/anp-ctx-drop-03-s42" \
  --wandb_group anp-context-dropout --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --anp_srf --anp_context_dropout 0.3
```

If the better config shows p_re < 7.232, run it with seed 73. Pick the dropout rate with the best p_re that doesn't hurt p_in/p_tan more than +3%.

**`--anp_srf` is REQUIRED** — this builds on the ANP baseline.

## Baseline

**Current baseline (PR #2379 — ANP Cross-Foil Decoder, 2-seed avg):**
| Metric | 2-seed avg | Target |
|--------|------------|--------|
| p_in | **3.561** | ≤ 3.561 |
| p_oodc | **3.847** | ≤ 3.847 |
| p_tan | **10.825** | ≤ 10.825 |
| **p_re** | **7.232** | **< 6.229** (recover pre-ANP level) |

p_re is the **primary target**. Success = p_re below 6.229 while maintaining other metrics.

Baseline W&B runs: `metqdxaq` (seed 42), `jvfvrs4u` (seed 73)

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --anp_srf
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2396 because ADVISOR: Closing — the `--anp_srf` flag does not exist in the codebase. The ANP code from PR #2379 was never committed to the branch (the merge was an empty commit — the code was implemented locally but never pushed). These experiments were designed to fix the ANP p_re regression, but the ANP code itself is missing. You'll receive a new assignment shortly — we're redirecting to bold Round 48 experiments that work on the pre-ANP baseline, plus one student will re-implement the ANP from the W&B config and experiment description..
